sharedbuffers/mapped_struct.py

```diff
@@ -4394,7 +4628,7 @@ def get_packer(self, obj):
             ]))
```

Comment:
I'm getting these errors in tests:

```
======================================================================
ERROR: testUnpack (tests.pool.SmallIntContainerPackingTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/claudiofreire/src/sharedbuffers/tests/pool.py", line 114, in testUnpack
    pos = p.pack(self.schema, x)[0]
  File "/home/claudiofreire/src/sharedbuffers/sharedbuffers/pool.py", line 225, in pack
    pos = self._pack_into(schema, obj, section)
  File "/home/claudiofreire/src/sharedbuffers/sharedbuffers/pool.py", line 171, in _pack_into
    buf = schema.pack(obj, idmap, implicit_offs=implicit_offs)
  File "sharedbuffers/mapped_struct.py", line 4809, in sharedbuffers.mapped_struct.Schema.pack
    def pack(self, obj, idmap = None, packer = None, padding = None, implicit_offs = 0):
  File "sharedbuffers/mapped_struct.py", line 4819, in sharedbuffers.mapped_struct.Schema.pack
    endp = self.pack_into(obj, buf, 0, idmap, packer, padding, implicit_offs)
  File "sharedbuffers/mapped_struct.py", line 4774, in sharedbuffers.mapped_struct.Schema.pack_into
    packer, padding = self.get_packer(obj)
  File "sharedbuffers/mapped_struct.py", line 4622, in sharedbuffers.mapped_struct.Schema.get_packer
    packer = struct.Struct("".join([
TypeError: sequence item 0: expected str instance, bytes found
```

Seems Struct.format is bytes, and the join there is using strings. Changing it to `b"".join` should fix it. There's quite a few places like these.

Reply:
Yep, seems 3.6 has bytes in Struct.format, but not the other versions. Just doing b"".join doesn't fix it for all, it breaks in 3.7+.

Checking if there's a way that's worth the trouble. Otherwise we may have to drop support for 3.6-

Reply:
Ok, it looks ugly, but this seems to work, and it's a small effort to get 3.6 support as it's only in one place, so might as well do it:

``` diff
diff --git a/sharedbuffers/mapped_struct.py b/sharedbuffers/mapped_struct.py
index 41f19ce..51f4df5 100644
--- a/sharedbuffers/mapped_struct.py
+++ b/sharedbuffers/mapped_struct.py
@@ -4618,9 +4618,12 @@ class Schema(object):
             packer_key = present_bitmap
 
         rv = self.packer_cache.get(packer_key)
+        bitmap_packer_format = self.bitmap_packer.format
+        if isinstance(bitmap_packer_format, bytes):
+            bitmap_packer_format = bitmap_packer_format.decode('ascii')
         if rv is None:
             packer = struct.Struct("".join([
-                self.bitmap_packer.format,
+                bitmap_packer_format,
             ] + [
                 self.slot_struct_types[slot]
                 for i,slot in enumerate(self.slot_keys)
```

Reply:
Yep, I was just doing something like this